feat: support wildcard syntax in resource-strategy-fit plugin

ditingdapeng opened the pull request:
/kind feature

## What this PR does / why we need it:

This PR implements wildcard syntax support for the resource-strategy-fit plugin to address scalability issues in GPU resource configuration.

**Problem**: Current configuration requires specifying each GPU type individually (e.g., `cloudml.gpu/r49-24g`, `cloudml.gpu/r39-24g`), creating operational burden in diverse GPU environments.

**Solution**: 
- Add wildcard pattern validation for resource configurations
- Support prefix matching with trailing asterisk (e.g., 'cloudml.gpu/*')
- Filter out invalid patterns like single '*' or multiple asterisks
- Add comprehensive test cases for wildcard functionality

## Which issue(s) this PR fixes:
Fixes #4449

## Special notes for your reviewer:
- Implementation follows "Solution 1: Wildcard Syntax Support" from the issue discussion
- All existing tests pass, new tests validate wildcard functionality
- Warning logs help identify invalid configurations
- Maintains backward compatibility

## Does this PR introduce a user-facing change?
```release-note
resource-strategy-fit plugin now supports wildcard syntax (e.g., 'cloudml.gpu/*') for scalable GPU resource configuration
```

ditingdapeng commented on pkg/scheduler/plugins/resource-strategy-fit/resource_strategy_fit.go:
```diff
@@ -137,6 +151,48 @@ func (rsf *resourceStrategyFitPlugin) OnSessionOpen(ssn *framework.Session) {
+		return false
+	}
+
+	asteriskCount := strings.Count(pattern, "*")
+	if asteriskCount != 1 {
+		return false
+	}
+
+	lastIndex := strings.LastIndex(pattern, "*")
+	return lastIndex == len(pattern)-1
```
Thank you for the review and suggestion.

While the logic `strings.Count(pattern, "*") == 1` would indeed ensure there's only one asterisk, it doesn't guarantee that the asterisk is at the **end** of the string, which is crucial for our prefix matching implementation.

Consider these examples:
- `"*gpu"` - has exactly 1 asterisk but it's at the beginning
- `"gpu*memory"` - has exactly 1 asterisk but it's in the middle
- `"gpu*"` - has exactly 1 asterisk and it's at the end ✓

Our current implementation specifically supports **suffix wildcard matching** (prefix matching with trailing asterisk) as mentioned in the issue discussion. The position check `lastIndex == len(pattern)-1` is essential to ensure we only accept patterns like `"cloudml.gpu*"` and reject patterns like `"*gpu"` or `"gpu*memory"`.

The current validation logic is intentionally designed this way to:
1. Ensure exactly one asterisk exists
2. Ensure that asterisk is at the end (enabling prefix matching)
3. Maintain clear semantics for the wildcard feature

If we want to optimize, we could use `strings.HasSuffix(pattern, "*")` instead of `strings.LastIndex`, but the position validation itself should remain.

JesseStutler replied:
The meaning of gemini is like this:
```
    if !strings.HasSuffix(pattern, "*") {
		return false
	}

	return strings.Count(pattern, "*") == 1
```
This already ensures that * needs to be at the end and can only have one

JesseStutler commented:
/ok-to-test
/cc

JesseStutler commented on pkg/scheduler/plugins/resource-strategy-fit/resource_strategy_fit.go:
```diff
@@ -104,6 +105,19 @@ func calculateWeight(args framework.Arguments) ResourceStrategyFit {
 	for k, v := range resources {
+		configKeyStr := string(k)
+
+		if strings.Contains(configKeyStr, "*") {
```
Why do we need to add a `strings.Contains` check in the outer layer? Shouldn't this be put in `isValidWildcardPattern`?

JesseStutler commented on pkg/scheduler/plugins/resource-strategy-fit/resource_strategy_fit.go:
```diff
@@ -137,6 +151,48 @@ func (rsf *resourceStrategyFitPlugin) OnSessionOpen(ssn *framework.Session) {
+
+	for configKey, config := range resources {
+		configKeyStr := string(configKey)
+		if strings.HasSuffix(configKeyStr, "*") {
```
The previous one has been verified by `isValidWildcardPattern`, so there is no need to add * here. We should put all the format validation logic in `isValidWildcardPattern`

JesseStutler commented on pkg/scheduler/plugins/resource-strategy-fit/resource_strategy_fit.go:
```diff
@@ -137,6 +151,48 @@ func (rsf *resourceStrategyFitPlugin) OnSessionOpen(ssn *framework.Session) {
+}
+
+func findResourceConfigWithPrefix(resourceName string, resources map[v1.ResourceName]ResourcesType) (ResourcesType, bool) {
+	if config, exists := resources[v1.ResourceName(resourceName)]; exists {
```
A comment can be added here like if exactMatch hits, exactMatch will prevail

JesseStutler commented on docs/design/resource-strategy-fit-scheduling.md:
```diff
@@ -72,6 +75,76 @@ node score:
+    - Asterisk at the beginning (`*/gpu`)
+    - Multiple asterisks (`vendor.com/**`)
+
+**Matching Priority:**
```
This part can move to syntax rules

ditingdapeng replied:
done

ditingdapeng commented:
@JesseStutler  Thanks for the feedback! I'll go through each comment and make revisions over the weekend, then submit.

ditingdapeng commented:
@JesseStutler  Hello, I have addressed all the review comments. Thank you for taking the time to review!

## Comment 1: Simplification of isValidWildcardPattern function
Your suggestion: "This already ensures that * needs to be at the end and can only have one"

My changes: ✅ Addressed Adopted your suggestion and simplified the original complex logic to:

```
if !strings.HasSuffix(pattern, "*") {
    return false
}
return strings.Count(pattern, "*") == 1
```

## Comment 2: Question about outer strings.Contains check
Your suggestion: "Why do we need to add a strings.Contains check in the outer layer? Shouldn't this be put in isValidWildcardPattern?"

My changes: ✅ Addressed Kept the strings.Contains(configKeyStr, "*") check for detection purposes, which is separate from the validation responsibility of isValidWildcardPattern . This follows the single responsibility principle.

## Comment 3: Removing redundant wildcard checks
Your suggestion: "The previous one has been verified by isValidWildcardPattern, so there is no need to add * here. We should put all the format validation logic in isValidWildcardPattern"

My changes: ✅ Addressed Removed the original strings.HasSuffix(configKeyStr, "*") check because:

- All format validation is now handled in isValidWildcardPattern
- Only need strings.Contains(configKeyStr, "*") here to detect wildcard patterns

## Comment 4: Documentation and wildcard-related comments
Your suggestions: "I think keep ###Configuration Guidelines and ### Risk Analysis are enough." and "A comment can be added here like if exactMatch hits, exactMatch will prevail"

My changes: ✅ Addressed Updated the documentation structure and added code comments as suggested.

ditingdapeng commented:
Hi @JesseStutler , I've just pushed the updates. This branch is now rebased on the latest `master` and the previous issues have been fixed.

Could you please take another look when you have a moment? Thanks!

Monokaix commented on docs/design/resource-strategy-fit-scheduling.md:
```diff
@@ -72,6 +75,61 @@ node score:
+  nvidia.com/gpu-v100:     # Exact match - highest priority
+    type: MostAllocated
+    weight: 3
+  nvidia.com/gpu/*:        # Wildcard match - covers other NVIDIA GPUs
```
How many match pattern will be supported, only wildcard? Do we need to  add other patterns or we only support one pattern?

Monokaix commented on pkg/scheduler/plugins/resource-strategy-fit/resource_strategy_fit.go:
```diff
@@ -137,6 +149,52 @@ func (rsf *resourceStrategyFitPlugin) OnSessionOpen(ssn *framework.Session) {
+
+	return bestConfig, found
+}
 func Score(task *api.TaskInfo, node *api.NodeInfo, weight ResourceStrategyFit) float64 {
```
Please add a blank line.

ditingdapeng commented on pkg/scheduler/plugins/resource-strategy-fit/resource_strategy_fit.go:
```diff
@@ -137,6 +149,52 @@ func (rsf *resourceStrategyFitPlugin) OnSessionOpen(ssn *framework.Session) {
+	for configKey, config := range resources {
+		configKeyStr := string(configKey)
+		// All wildcard patterns have been validated by isValidWildcardPattern
+		if strings.Contains(configKeyStr, "*") {
```
I considered using regexp but chose the simple wildcard approach for these reasons:

1. **Simplicity and Performance**: Simple string operations like `strings.HasPrefix()` are much faster than regexp compilation and matching.

2. **User Experience**: Wildcard syntax is more intuitive for K8s administrators, similar to RBAC rules patterns.

3. **Security**: Limiting to simple patterns reduces regex injection risks and performance issues.

4. **Scope Alignment**: The primary use case is matching GPU resources like `nvidia.com/gpu-*` variants, which wildcard matching handles perfectly.

For example, with wildcard `nvidia.com/gpu/*`, we can match resources like:
- `nvidia.com/gpu/tesla-v100`
- `nvidia.com/gpu/rtx-3080` 
- `nvidia.com/gpu/a100`

This covers the common GPU resource naming patterns without regex complexity.

ditingdapeng commented on docs/design/resource-strategy-fit-scheduling.md:
````diff
@@ -72,6 +75,76 @@ node score:
+    weight: 1
+```
+
+**Implementation Details:**
````
done

Monokaix commented:
/approve
Thanks!

ditingdapeng commented:
@Thor-wl  @alcorj-mizar 
Hello, Could you please help review this PR when you have time? Thanks!

kingeasternsun commented:
> @Thor-wl @alcorj-mizar Hello, Could you please help review this PR when you have time? Thanks!

Please squash you commits to one commit.

kingeasternsun commented:
/lgtm
